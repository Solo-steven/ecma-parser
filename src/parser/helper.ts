import { SyntaxKinds } from "@/src/syntax/kinds";

export function getBinaryPrecedence(kind: SyntaxKinds) {
    switch(kind) {
        case SyntaxKinds.LogicalOROperator:
            return 4;
        case SyntaxKinds.LogicalANDOperator:
            return 5;
        case SyntaxKinds.BitwiseOROperator:
            return 6;
        case SyntaxKinds.BitwiseXOROperator:
            return 7;
        case SyntaxKinds.BitwiseANDOperator:
            return 8;
        case SyntaxKinds.StrictEqOperator:
        case SyntaxKinds.StrictNotEqOperator:
        case SyntaxKinds.EqOperator:
        case SyntaxKinds.NotEqOperator:
            return 9;
        case SyntaxKinds.InKeyword:
        case SyntaxKinds.InstanceofKeyword:
        case SyntaxKinds.GtOperator:
        case SyntaxKinds.GeqtOperator:
        case SyntaxKinds.LeqtOperator:
        case SyntaxKinds.LtOperator:
            return 10;
        case SyntaxKinds.BitwiseLeftShiftOperator:
        case SyntaxKinds.BitwiseRightShiftOperator:
        case SyntaxKinds.BitwiseRightShiftFillOperator:
            return 11;
        case SyntaxKinds.PlusOperator:
        case SyntaxKinds.MinusOperator:
            return 12;
        case SyntaxKinds.ModOperator:
        case SyntaxKinds.DivideOperator:
        case SyntaxKinds.MultiplyOperator:
            return 13;
        case SyntaxKinds.ExponOperator:
            return 14;
        default:
            return -1;
    }
}

export function isBinaryOps(kind: SyntaxKinds) {
    return getBinaryPrecedence(kind) > 0;
}