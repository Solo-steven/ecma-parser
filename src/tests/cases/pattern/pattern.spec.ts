import path from "path";
import { readFileAsString, toAST } from "../../helper";

describe("cases/ destructure pattern", () => {
    test("nested object pattern used in declaration", async () => {
        const code = await readFileAsString(path.join(__dirname, "nestObjectPatternDeclaration.js"));
        expect(toAST(code)).toMatchSnapshot();
    });
    test("array pattern in declaration", async () => {
        const code = await readFileAsString(path.join(__dirname, "arrayPatternInDeclaration.js"));
        expect(toAST(code)).toMatchSnapshot();
    });
    test("object pattern in declaration", async () => {
        const code = await readFileAsString(path.join(__dirname, "objectPatternInDeclaration.js"));
        expect(toAST(code)).toMatchSnapshot();
    });
    test("nested array pattern in declaration", async () => {
        const code = await readFileAsString(path.join(__dirname, "nestedArrayPatternInDeclaration.js"));
        expect(toAST(code)).toMatchSnapshot();
    });
})