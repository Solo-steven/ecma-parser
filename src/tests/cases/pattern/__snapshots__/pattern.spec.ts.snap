// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cases/ destructure pattern array pattern in declaration 1`] = `
{
  "body": [
    {
      "declarations": [
        {
          "id": {
            "elements": [
              {
                "kind": 10117,
                "name": "state",
              },
              {
                "kind": 10117,
                "name": "setState",
              },
            ],
            "kind": 10178,
          },
          "init": {
            "arguments": [],
            "callee": {
              "kind": 10117,
              "name": "useState",
            },
            "kind": 10167,
            "optional": false,
          },
          "kind": 10140,
        },
      ],
      "kind": 10139,
      "variant": "const",
    },
  ],
  "kind": 10120,
}
`;

exports[`cases/ destructure pattern object pattern in declaration 1`] = `
{
  "body": [
    {
      "declarations": [
        {
          "id": {
            "kind": 10176,
            "properties": [
              {
                "computed": false,
                "key": {
                  "kind": 10117,
                  "name": "data",
                },
                "kind": 10177,
                "shorted": false,
                "value": undefined,
              },
              {
                "computed": false,
                "key": {
                  "kind": 10117,
                  "name": "name",
                },
                "kind": 10177,
                "shorted": false,
                "value": undefined,
              },
              {
                "computed": false,
                "key": {
                  "kind": 10117,
                  "name": "value",
                },
                "kind": 10177,
                "shorted": false,
                "value": undefined,
              },
            ],
          },
          "init": {
            "kind": 10117,
            "name": "state",
          },
          "kind": 10140,
        },
      ],
      "kind": 10139,
      "variant": "const",
    },
  ],
  "kind": 10120,
}
`;
